cpp/src/arrow/io/file.cc

```diff
@@ -526,13 +588,22 @@ Status MemoryMappedFile::Write(const void* data, int64_t nbytes) {
   if (!memory_map_->opened() || !memory_map_->writable()) {
     return Status::IOError("Unable to write");
   }
-  if (nbytes + memory_map_->position() > memory_map_->size()) {
-    return Status::Invalid("Cannot write past end of memory map");
-  }
   return WriteInternal(data, nbytes);
 }
 
+Status MemoryMappedFile::Resize(int64_t new_size) {
+  if (new_size >= memory_map_->capacity()) {
+    return memory_map_->Reserve(new_size);
+  } else {
+    return memory_map_->Resize(new_size, /*shrink_to_fit*/ true);
+  }
+}
+
 Status MemoryMappedFile::WriteInternal(const void* data, int64_t nbytes) {
+  if (nbytes + memory_map_->position() > memory_map_->size()) {
+    int64_t new_size = BitUtil::NextPower2(nbytes + memory_map_->position());
```

Comment:
Hmm, really? I would never expect a file to over-allocate memory / disk space for me. I'm already surprised that `Write()` on a memory-mapped file would append silently. For comparison, in Python I can't implicitly write past the end of a mmap'ed file, I have to extend it explicitly:

```python
>>> f = open('FOO', 'w+')
>>> f.truncate(4096)
4096
>>> m = mmap.mmap(f.fileno(), 4096)
>>> m.write(b'x'*8000)
Traceback (most recent call last):
  File "<ipython-input-12-183b7a903b2b>", line 1, in <module>
    m.write(b'x'*8000)
ValueError: data out of range
>>> m.resize(8192)
>>> m.write(b'x'*8000)
8000
```

Reply:
Design decisions cont: initial reason for me to implement mmap resizing was that I wanted to use the `RecordBatchFileWriter` as simply as the `RecordBatchStreamWriter` with a `BufferOutputStream` and a resizable buffer. I reasoned that if we can grow a buffer exponentially in RAM (as in case of the `PoolBuffer`), we might as well do it on disk. 

I see your and Wes' point though, that such a behaviour might be surprising for the user. An alternative would be a transparent option to define the resizing policy (let's say a `SetGrowthPolicy` method is options `None` (default) / `ToNextPage` / `ConstNBytes` / `ConstFactor` etc). A possible implementation would be to add a new interface `ResizableFile`, that way we can extend other file types with this functionality, too.

Reply:
If this change needs to add a resizing policy, I think it would be first be worthwhile showcasing a significant performance improvement (to avoid adding complexity and a maintenance burden that isn't ultimately warranted).

`ResizableFile` is indeed another possible approach... though we may not necessarily make it a separate class, and instead have `Resize` return a `NotImplemented` error by default?

(I'm not sure what's our policy on this, but the proliferation of abstract base classes doesn't necessarily sound very compelling)